flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/ElasticsearchSinkBase.java

```diff
@@ -211,6 +283,23 @@ public void invoke(T value) throws Exception {
 	}
 
 	@Override
+	public void initializeState(FunctionInitializationContext context) throws Exception {
+		// no initialization needed
+	}
+
+	@Override
+	public void snapshotState(FunctionSnapshotContext context) throws Exception {
+		checkErrorAndRethrow();
+
+		if (flushOnCheckpoint) {
+			do {
+				bulkProcessor.flush();
```

Comment:
Ah, I see the problem here ...
The bulk processor's internal `bulkRequest.numberOfActions() == 0` will become `true` as soon as it starts executing the flush, and not after `afterBulk` is invoked.

So, since our `numPendingRequests` implementation relies on the `afterBulk` callback, we might have busy loops on `bulkProcessor.flush()` while we wait for `numPendingRequests` to become 0.

This is quite a nice catch actually! So no worries on bringing it up now.